atomic_reactor/plugins/pre_resolve_composes.py

```diff
@@ -258,19 +267,16 @@ def resolve_signing_intent(self):
         self.compose_config.set_signing_intent(signing_intent['name'])
 
     def forward_composes(self):
-        repos_by_arch = {}
+        repos_by_arch = defaultdict(list)
         # set overrides by arch if arches are available
         for compose_info in self.composes_info:
-            for arch in compose_info.get('arches') or []:
-                repos_by_arch.setdefault(arch, [])
+            for arch in compose_info.get('arches', '').split(" "):
                 repos_by_arch[arch].append(compose_info['result_repofile'])
+            if not compose_info.get('arches'):
+                repos_by_arch[None].append(compose_info['result_repofile'])
 
-        for arch in repos_by_arch:
-            override_build_kwarg(self.workflow, 'yum_repourls', repos_by_arch[arch], arch)
-        # set generic overrrides otherwise
-        else:
-            yum_repourls = [compose_info['result_repofile'] for compose_info in self.composes_info]
-            override_build_kwarg(self.workflow, 'yum_repourls', yum_repourls, None)
+        for arch, repofiles in repos_by_arch.items():
+            override_build_kwarg(self.workflow, 'yum_repourls', repofiles, arch)
```

Comment:
If we ended up setting `None`, those repos will get lost. We need to distribute overrides from `None` to all the actual architectures if we want them to get used:
```
try:
    noarch_repos = repos_by_arch.pop(None)
except KeyError:
    pass
else:
    for repos in repos_by_arch.values():
        repos.extend(noarch_repos)
for arch, repofiles in repos_by_arch.items():
    override_build_kwarg(...)
```

But first: do we really need to handle the case where a compose response does not tell us which arches it is for?

Reply:
What? No, I don't think that's right. override(None) sets an override for all arches, then override(arch) sets it for an arch.

we should do the pop to guarantee order, but we don't need to distribute because that's already handled by override_build_kwarg.

Reply:
The way `orchestrate_build` plugin reads this parameter it does not merge `None` and `<platform>` keys. If both are set, `<platform>` will override `None`. That's because the internal story for adding that behavior called explicitly to not merge values recursively.

That means they need to be merged here, and `override_build_kwarg(...., None)` should never be called from this `resolve_composes` plugin.

But yes, this may not be needed at all. Talked to Jan from ODCS and he thinks arches "will always be set to something".

Reply:
okay.

Reply:
I'm confused now. Why doesn't Tim's code fail to set any repos when None is the only entry in repos_by_arch?

Reply:
Perhaps it should be only be distributed if there are non-None repos_by_arch:
```
if any(arch != None for arch in repos_by_arch):
    try:
        noarch_repos = repos_by_arch.pop(None)
    except KeyError:
        pass
    else:
        for repos in repos_by_arch.values():
            repos.extend(noarch_repos)
        for arch, repofiles in repos_by_arch.items():
            override_build_kwarg(...)
```

Reply:
Because `pop` will remove the `None` key and its value from `repos_by_arch`, leaving it an empty dict which is still iterable:

```python
>>> {}.values()
[]
```

Oh! I think you're pointing out that the code will essentially be a noop in this case!

Maybe it needs this to always call `override_build_kwarg(...., noarch_repos, None)` then.

Reply:
@lcarva, in other words, you think it should not `pop()` it?
```
try:
    noarch_repos = repos_by_arch[None]
except KeyError:
    pass
else:
    # Copy the repos relevant for all architectures
    # into each arch-specific override
    for repos in repos_by_arch.values():
        repos.extend(noarch_repos)
for arch, repofiles in repos_by_arch.items():
    override_build_kwarg(...)
```

Reply:
We should pop it, distribute it we can, and set None if there's nothing left in repos_by_arch.

Fixed.